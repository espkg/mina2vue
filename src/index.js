import parser from '@babel/parser'
import generator from '@babel/generator'
import WXMLParser from './wxml-parser.js'
import templateTransformer from './template-transformer.js'
import minaScriptTransformer from './mina-script-transformer'
import MFScriptTransformer from './mf-script-transformer'
import vueGenerator from './vue-generator'
import styleTransformer from './style-transformer'
import HTMLGenerator from './html-generator'

const wxmlParser = new WXMLParser()

async function handleWXML (templateStr) {
  const templateAST = await wxmlParser.parse(templateStr)
  const distAST = await templateTransformer(templateAST)
  HTMLGenerator.configure({
    disableAttribEscape: true
  })
  const distCode = HTMLGenerator(distAST)

  return distCode
}

function handleWXSS (styleStr) {
  return styleTransformer(styleStr)
}

function handleJS (scriptStr) {
  const scriptAST = parser.parse(scriptStr, {
    sourceType: 'module'
  })
  let distAST = minaScriptTransformer(scriptAST)
  distAST = MFScriptTransformer(distAST)
  const distCode = generator.default(distAST).code

  return distCode
}

export default async function YG (template, style, scripts) {
  const distTemplate = await handleWXML(template)
  return vueGenerator(
    distTemplate,
    handleWXSS(style),
    handleJS(scripts)
  )
}
