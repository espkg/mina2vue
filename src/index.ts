import wxmlConverter from './language-wxml'
import wxssConverter from './language-wxss'
import jsConverter from './language-js/index'
import vueGenerator from './language-vue/generator'

export {
  jsConverter,
  wxmlConverter,
  wxssConverter,
  vueGenerator
}

// export default function (sourceCodes : string, options : any) {

// }
