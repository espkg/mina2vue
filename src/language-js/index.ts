import parser from './parser'
import transformer from './transformation'
import generator from './generator'

// TODO: 默认 options
export default function (sourceCode : string, options : any = { type: 'page' }) : string {
  const ast : babel.types.File = parser(sourceCode)

  const distAst = transformer(ast, options)

  const distCode : string = generator(distAst)

  return distCode
}
