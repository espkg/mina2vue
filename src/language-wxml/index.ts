import parser from './parser'
import transformer from './transformer'
import generator from './generator'

// TODO: 默认 options
export default async function (sourceCode : string, options : any = { }) {
  const ast = await parser(sourceCode)

  const distAst = await transformer(ast, options)

  generator.configure({
    disableAttribEscape: true
  })
  const distCode : string = generator(distAst)

  // Undo the hacky fix from parser
  return distCode.replace(/<>/g, (m) => {
    return `<`
  })
}
