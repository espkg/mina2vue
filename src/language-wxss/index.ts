import transformer from './transformer'

// TODO: 默认 options
export default function (sourceCode : string, options : any = { type: 'page' }) : string {
  const distCode = transformer(sourceCode, options)

  return distCode
}
