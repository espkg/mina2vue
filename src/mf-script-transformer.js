import t from '@babel/types'
import traverseModule from '@babel/traverse'
const traverse = traverseModule.default

const H5ActivityModuleMap = {
  MFPAGE: '@/components/base/mf-page'
}
const needAddMixin = new Set([])
const alreadyAddMixin = new Set([])

export default function MFScriptTransformer (ast) {
  traverse(ast, {
    ImportDeclaration (path) {
      const { node } = path
      const { specifiers } = node
      // 寻找 H5 公共包，未来改成用 get 方法实现，比较优雅
      // console.log(path.get('MFPage'))
      for (let i = 0; i < specifiers.length; i++) {
        const specifier = specifiers[i]
        const moduleName = specifier.local.name
        if (H5ActivityModuleMap[moduleName.toUpperCase()]) {
          node.source = t.stringLiteral(H5ActivityModuleMap[moduleName.toUpperCase()])
        }
      }
    },
    CallExpression (path) {
      const { node } = path
      if (t.isMemberExpression(node.callee)) {
        const { callee } = node
        if (t.isThisExpression(callee.object) && callee.property.name === 'hideMFLoading') {
          // 需判断是否已经添加，或重复添加
          needAddMixin.add('MFPageMixin')
        }
      }
    }
  })

  traverse(ast, {
    ImportDeclaration (path) {
      if (needAddMixin.has('MFPageMixin') && !alreadyAddMixin.has('MFPageMixin')) {
        path.insertAfter([
          t.importDeclaration([
            t.importDefaultSpecifier(t.identifier('MFPageMixin'))
          ], t.stringLiteral('@/mixins/MFPage.mixins'))
        ])
        alreadyAddMixin.add('MFPageMixin')
      }
    },
    ExportDefaultDeclaration (path) {
      const { node } = path
      const elements = Array.from(needAddMixin).map((mixin) => {
        return t.identifier(mixin)
      })
      node.declaration.properties.push(
        t.objectProperty(t.identifier('mixins'), t.arrayExpression(elements))
      )
    }
  })

  return ast
}
