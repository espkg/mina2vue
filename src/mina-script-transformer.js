import t from '@babel/types'
import traverseModule from '@babel/traverse'
const traverse = traverseModule.default

const watchMethods = t.objectProperty(
  t.identifier('watch'),
  t.objectExpression([])
)

function solveOtherProps (parent, prop) {
  parent.properties.push(prop)
}

function solveData (parent, prop) {
  const oldData = prop.value.properties
  const vueData = t.objectProperty(
    t.identifier('data'),
    t.arrowFunctionExpression([], t.objectExpression(oldData))
  )
  parent.properties.push(vueData)
}

function solveProps (parent, prop) {
  const oldProps = prop.value.properties

  oldProps.forEach((p) => {
    const pProps = p.value.properties
    pProps && pProps.forEach((pProp, i) => {
      if (pProp.key.name === 'observer') {
        // observer 转 watch
        watchMethods.value.properties.push(
          t.objectMethod('method',
            p.key,
            pProp.params,
            pProp.body
          )
        )
        pProps.splice(i, i)
      }
    })
  })

  const vueProps = t.objectProperty(
    t.identifier('props'),
    t.objectExpression(oldProps)
  )

  parent.properties.push(vueProps)
}

export default function minaScriptTransformer (ast) {
  traverse(ast, {
    CallExpression (path) {
      if (path.node.callee.name === 'MFPage') {
        const rootProps = path.node.arguments[0].properties
        const vueProps = t.objectExpression([])
        const vueMethods = t.objectExpression([])

        rootProps.forEach((prop) => {
          if (t.isObjectMethod(prop) || (t.isObjectProperty(prop) && t.isFunctionExpression(prop.value))) {
            // 成员方法
            switch (prop.key.name) {
              case 'onLoad':
                vueProps.properties.push(t.objectMethod(
                  'method',
                  t.identifier('mounted'),
                  prop.value.params,
                  prop.value.body
                ))
                break
              case 'onReady':
                vueProps.properties.push(t.objectMethod(
                  'method',
                  t.identifier('created'),
                  prop.value.params,
                  prop.value.body
                ))
                break
              case 'onShow':
                // solveReady()
                break
              case 'onHide':
                // solveReady()
                break
              case 'onUnload':
                vueProps.properties.push(t.objectMethod(
                  'method',
                  t.identifier('destroyed'),
                  prop.value.params,
                  prop.value.body
                ))
                break
              case 'onShareAppMessage':
                // solveReady()
                break
              default:
                vueMethods.properties.push(prop)
            }
          } else if (t.isObjectProperty(prop)) {
            // 成员属性
            switch (prop.key.name) {
              case 'data':
                solveData(vueProps, prop)
                break

              case 'properties':
                solveProps(vueProps, prop)
                break

              default:
                solveOtherProps(vueProps, prop)
            }
          }
        })

        // 插入 methods
        vueProps.properties.push(t.objectProperty(
          t.identifier('methods'),
          vueMethods
        ))

        // 插入 watch
        vueProps.properties.push(watchMethods)

        path.insertAfter(
          t.exportDefaultDeclaration(vueProps)
        )

        path.remove()
      }
    }
  })

  traverse(ast, {
    CallExpression (path) {
      const { node } = path
      // this.setData -> this.x = xxx
      if (t.isThisExpression(node.callee.object) && node.callee.property.name === 'setData') {
        // 如果 setData 有第二个参数
        if (node.arguments.length === 2) {
          const afterFunction = node.arguments[1]
          path.insertAfter(afterFunction.body.body)
        }

        const props = node.arguments[0].properties.reverse() // 倒序插入
        props.forEach((prop) => {
          path.insertAfter(
            t.assignmentExpression('=', t.memberExpression(t.thisExpression(), prop.key), prop.value)
          )
        })

        path.remove()
      }
    },
    MemberExpression (path) {
      const { node } = path
      // this.data.foo -> this.foo
      if (t.isMemberExpression(node.object) &&
        t.isThisExpression(node.object.object) &&
        node.object.property.name === 'data'
      ) {
        path.replaceWith(
          t.memberExpression(
            t.thisExpression(),
            t.identifier(node.property.name)
          )
        )
      }
    }
  })

  return ast
}
