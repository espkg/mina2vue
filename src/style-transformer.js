export default function styleTransformer (sourceCode) {
  const reg = /([\d\s]+)*rpx/g
  const dpx = 0.5
  let res
  let css = sourceCode
  while ((res = reg.exec(css))) {
    css = res.input
    const index = res.index
    const rpxPix = res[0]
    const num = rpxPix.replace('rpx', '').trim().replace(/(^:)|(:$)/g, '')
    const remPix = `${dpx * num}px`
    css = `${css.slice(0, index)} ${remPix}${css.slice(index + rpxPix.length, css.length)}`
  }

  return css
}
