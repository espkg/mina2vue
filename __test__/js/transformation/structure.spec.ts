import { jsConverter } from '../../../src'

it('should output vue structure', () => {
  const input = `Page({
  })`
  const output = jsConverter(input)

  expect(output).toMatchSnapshot()
})

it('should output component to vue structure', () => {
  const input = `Component({
  })`
  const output = jsConverter(input, {
    type: 'component'
  })

  expect(output).toMatchSnapshot()
})

it('should convert components\' properties to vue props', () => {
  const input = `Component({
    properties: {}
  })`
  const output = jsConverter(input, {
    type: 'component'
  })

  expect(output).toMatchSnapshot()
})

it('convert simple property definition', () => {
  const input = `Component({
    properties: {
      foo1: Boolean
    }
  })`
  const output = jsConverter(input, {
    type: 'component'
  })

  expect(output).toMatchSnapshot()
})

it('convert property definition with default value', () => {
  const input = `Component({
    properties: {
      foo2: {
        type: Number,
        value: 0
      }
    }
  })`
  const output = jsConverter(input, {
    type: 'component'
  })

  expect(output).toMatchSnapshot()
})

it('convert property definition with default value on reference types', () => {
  const input = `Component({
    properties: {
      foo3: {
        type: Object,
        value: {
          bar: 1
        }
      },
      foo4: {
        type: Array,
        value: [1, 2, 3]
      }
    }
  })`
  const output = jsConverter(input, {
    type: 'component'
  })

  expect(output).toMatchSnapshot()
})

it('convert property definition with Observer', () => {
  const input = `Component({
    properties: {
      foo5: {
        type: Number,
        value: 0,
        observer: function (newVal, oldVal) {
          // 属性值变化时执行
        }
      }
    }
  })`
  const output = jsConverter(input, {
    type: 'component'
  })

  expect(output).toMatchSnapshot()
})

it('convert property definition with Observer on reference types', () => {
  const input = `Component({
    properties: {
      foo6: {
        type: Object,
        value: {
          bar: 1
        },
        observer: function (newVal, oldVal) {
          // 属性值变化时执行
        }
      }
    }
  })`
  const output = jsConverter(input, {
    type: 'component'
  })

  expect(output).toMatchSnapshot()
})

it('should avoid add additional watch property', () => {
  const input = `Component({
    properties: {
      foo7: {
        type: Number,
        value: 0,
        observer: function (newVal, oldVal) {
          // 属性值变化时执行
        }
      }
    },
    watch: {}
  })`
  const output = jsConverter(input, {
    type: 'component'
  })

  expect(output).toMatchSnapshot()
})

it('should insert a watch property after existing properties', () => {
  const input = `Component({
    properties: {
      foo8: {
        type: Number,
        value: 0,
        observer: function (newVal, oldVal) {
          // 属性值变化时执行
        }
      }
    },
    watch: {
      foo: () => {}
    }
  })`
  const output = jsConverter(input, {
    type: 'component'
  })

  expect(output).toMatchSnapshot()
})

it('should convert page data to vue format', () => {
  const input = `Page({
    data: {}
  })`
  const output = jsConverter(input)

  expect(output).toMatchSnapshot()
})

it('should convert component data to vue format', () => {
  const input = `Component({
    data: {}
  })`
  const output = jsConverter(input, {
    type: 'component'
  })

  expect(output).toMatchSnapshot()
})

it('convert lifecycle methods to vue name properly', () => {
  const input = `Component({
    behaviors: ['watch'],
    attached: () => {},
    detached: function () {},
    ready () {}
  })`
  const output = jsConverter(input, {
    type: 'component'
  })

  expect(output).toMatchSnapshot()
})

it('convert page lifecycle methods to vue name properly', () => {
  const input = `Page({
    onLoad: () => {},
    onReady: function () {},
    onUnload () {}
  })`
  const output = jsConverter(input)

  expect(output).toMatchSnapshot()
})

it('put other page methods & properties in vue methods property', () => {
  const input = `Page({
    _data: {},
    initPage: () => {},
    handleClick (e) {}
  })`
  const output = jsConverter(input)

  expect(output).toMatchSnapshot()
})

it('should put other methods & properties in vue methods properly', () => {
  const input = `Page({
    _data: {},
    initPage: () => {},
    methods: {},
    handleClick (e) {}
  })`
  const output = jsConverter(input)

  expect(output).toMatchSnapshot()
})
