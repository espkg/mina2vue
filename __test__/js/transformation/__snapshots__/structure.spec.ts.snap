// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convert lifecycle methods to vue name properly 1`] = `
"export default {
  mixins: ['watch'],

  beforeMount() {},

  destroyed() {},

  mounted() {}

};"
`;

exports[`convert page lifecycle methods to vue name properly 1`] = `
"export default {
  created() {},

  mounted() {},

  destroyed() {}

};"
`;

exports[`convert property definition with Observer 1`] = `
"export default {
  props: {
    foo5: {
      type: Number,
      default: 0
    }
  },
  watch: {
    foo5: (newVal, oldVal) => {// 属性值变化时执行
    }
  }
};"
`;

exports[`convert property definition with Observer on reference types 1`] = `
"export default {
  props: {
    foo6: {
      type: Object,
      default: () => ({
        bar: 1
      })
    }
  },
  watch: {
    foo6: {
      handler(newVal, oldVal) {// 属性值变化时执行
      },

      deep: true
    }
  }
};"
`;

exports[`convert property definition with default value 1`] = `
"export default {
  props: {
    foo2: {
      type: Number,
      default: 0
    }
  }
};"
`;

exports[`convert property definition with default value on reference types 1`] = `
"export default {
  props: {
    foo3: {
      type: Object,
      default: () => ({
        bar: 1
      })
    },
    foo4: {
      type: Array,
      default: () => [1, 2, 3]
    }
  }
};"
`;

exports[`convert simple property definition 1`] = `
"export default {
  props: {
    foo1: Boolean
  }
};"
`;

exports[`put other page methods & properties in vue methods property 1`] = `
"export default {
  _data: {},
  methods: {
    initPage: () => {},

    handleClick(e) {}

  }
};"
`;

exports[`should avoid add additional watch property 1`] = `
"export default {
  props: {
    foo7: {
      type: Number,
      default: 0
    }
  },
  watch: {
    foo7: (newVal, oldVal) => {// 属性值变化时执行
    }
  }
};"
`;

exports[`should convert component data to vue format 1`] = `
"export default {
  data: () => ({})
};"
`;

exports[`should convert components' properties to vue props 1`] = `
"export default {
  props: {}
};"
`;

exports[`should convert page data to vue format 1`] = `
"export default {
  data: () => ({})
};"
`;

exports[`should insert a watch property after existing properties 1`] = `
"export default {
  props: {
    foo8: {
      type: Number,
      default: 0
    }
  },
  watch: {
    foo: () => {},
    foo8: (newVal, oldVal) => {// 属性值变化时执行
    }
  }
};"
`;

exports[`should output component to vue structure 1`] = `"export default {};"`;

exports[`should output vue structure 1`] = `"export default {};"`;

exports[`should put other methods & properties in vue methods properly 1`] = `
"export default {
  _data: {},
  methods: {
    initPage: () => {},

    handleClick(e) {}

  }
};"
`;
